### DISCUSSION TITLE
When to use scalar vs scalar_one_or_none vs one_or_none

### DISCUSSION THREAD

**Discussion by ShmuelTreiger** [sqlalchemy/sqlalchemy#6707] [Usage Questions] [answered]
Kind of basic, but when should I use Select.scalar() vs Select.one_or_none() vs Select.scalar_one_or_none()?

Obviously if I want the object, I need one of the `.scalar`s. Is there a difference between them? And if I am indifferent as to whether I get a scalar vs a row, is there a performance hit either way?

**Comment by CaselIT** [comment, marked as answer]
Hi

> Is there a difference between them?

The documentation for them is
- https://docs.sqlalchemy.org/en/14/core/connections.html#sqlalchemy.engine.Result.scalar
- https://docs.sqlalchemy.org/en/14/core/connections.html#sqlalchemy.engine.Result.scalar
- https://docs.sqlalchemy.org/en/14/core/connections.html#sqlalchemy.engine.Result.scalar_one_or_none


Regarding performance you mean on the db site? If so the query executed is exactly the same in all cases, so no

**Reply by ShmuelTreiger** [reply]
Oops, I was looking at Query.scalar() which seems basically the same as Select.scalar_one_or_none()

Thanks!

**Reply by CaselIT** [reply]
when using the 2.0 style queries with the orm, the results are unified for both core and orm execution.
So when using the orm with a single entity the result object by default returns a list of rows with a single element that's are the orm instances loaded. That's why `scalar*` are useful to obtain a list of orm instance (or a single one or none)

**Reply by ShmuelTreiger** [reply]
Yup yup. I understand. In the process of converting my project from using Query objects to Select objects. Appreciate the help!

As an aside: I assume .scalar_one_or_none will be slightly slower than just .one_or_none since it has to convert it to an object?

**Reply by zzzeek** [reply]
both one_or_none and scalar_one_or_none, when loaded against a SELECT that is for ORM objects, will have an ORM object present in the result that's returned.  The difference is that one_or_none gives you this:

```py
    (MyObject(), )
```

whereas scalar_one_or_none() gives you this:

```py
   MyObject()
```

so as far as creating an object, it's the same.  as far as performance of one_or_none() vs. scalar_one_or_none(), , the latter will skip having to create the Row object (the tuple) so there's an extremely small performance benefit to scalar_one_or_none for this case, but it would be barely measurable for less than many tens of thousands of operations.

**Reply by ShmuelTreiger** [reply]
Got it. Thank you! Appreciate the explanation!
